utilities/transactions/lock/range/range_tree/range_tree_lock_manager.cc

```diff
@@ -0,0 +1,572 @@
+namespace ROCKSDB_NAMESPACE {
+
+/////////////////////////////////////////////////////////////////////////////
+// RangeTreeLockManager -  A Range Lock Manager that uses PerconaFT's
+// locktree library
+/////////////////////////////////////////////////////////////////////////////
+
+RangeLockManagerHandle* NewRangeLockManager(
+    std::shared_ptr<TransactionDBMutexFactory> mutex_factory) {
+  std::shared_ptr<TransactionDBMutexFactory> use_factory;
+
+  if (mutex_factory)
+    use_factory = mutex_factory;
+  else
+    use_factory.reset(new TransactionDBMutexFactoryImpl());
+
+  return new RangeTreeLockManager(use_factory);
+}
+
+static const char SUFFIX_INFIMUM = 0x0;
+static const char SUFFIX_SUPREMUM = 0x1;
+
+void serialize_endpoint(const Endpoint& endp, std::string* buf) {
+  buf->push_back(endp.inf_suffix ? SUFFIX_SUPREMUM : SUFFIX_INFIMUM);
+  buf->append(endp.slice.data(), endp.slice.size());
+}
+
+void deserialize_endpoint(const DBT* dbt, EndpointWithString* endp) {
+  assert(dbt->size >= 1);
+  const char* dbt_data = (const char*)dbt->data;
+  char suffix = dbt_data[0];
+  assert(suffix == SUFFIX_INFIMUM || suffix == SUFFIX_SUPREMUM);
+  endp->inf_suffix = (suffix == SUFFIX_SUPREMUM);
+  endp->slice.assign(dbt_data + 1, dbt->size - 1);
+}
+
+// Get a range lock on [start_key; end_key] range
+Status RangeTreeLockManager::TryLock(PessimisticTransaction* txn,
+                                     uint32_t column_family_id,
+                                     const Endpoint& start_endp,
+                                     const Endpoint& end_endp, Env*,
+                                     bool exclusive) {
+  toku::lock_request request;
+  request.create(mutex_factory_);
+  DBT start_key_dbt, end_key_dbt;
+
+  TEST_SYNC_POINT("RangeTreeLockManager::TryRangeLock:enter");
+  std::string start_key;
+  std::string end_key;
+  serialize_endpoint(start_endp, &start_key);
+  serialize_endpoint(end_endp, &end_key);
+
+  toku_fill_dbt(&start_key_dbt, start_key.data(), start_key.size());
+  toku_fill_dbt(&end_key_dbt, end_key.data(), end_key.size());
+
+  // Use the txnid as "extra" in the lock_request. Then, KillLockWait()
+  // will be able to use kill_waiter(txn_id) to kill the wait if needed
+  // TODO: KillLockWait is gone and we are no longer using
+  // locktree::kill_waiter call. Do we need this anymore?
+  TransactionID wait_txn_id = txn->GetID();
+
+  auto lt = get_locktree_by_cfid(column_family_id);
+
+  request.set(lt, (TXNID)txn, &start_key_dbt, &end_key_dbt,
+              exclusive ? toku::lock_request::WRITE : toku::lock_request::READ,
+              false /* not a big txn */, (void*)wait_txn_id);
+
+  // This is for "periodically wake up and check if the wait is killed" feature
+  // which we are not using.
+  uint64_t killed_time_msec = 0;
+  uint64_t wait_time_msec = txn->GetLockTimeout();
+
+  // convert microseconds to milliseconds
+  if (wait_time_msec != (uint64_t)-1)
+    wait_time_msec = (wait_time_msec + 500) / 1000;
```

Comment:
even with the `if` check, `wait_time_msec + 500` is still possible to overflow?

Reply:
In this operation, no.
The value is obtained like so
```
  uint64_t wait_time_msec = txn->GetLockTimeout();
```
where txn->GetLockTimeout() returns int64_t, a signed value. The only possible negative value (with highest bit set)  is "-1" which means "no timeout". 


But I do see an issue with `wait_time_msec=-1`.  We will pass that value to the `request.wait(wait_time_msec, ...`) call.
`lock_request::wait` has no handling for the case where its `wait_time_ms` parameter is `(unt64_t)-1`: it immediately computes:

```
  uint64_t t_end = t_start + wait_time_ms * 1000;
```
which causes an overflow. This in turn, will cause lock that was acquired immediately to be counted as a "long wait".

I think I can fix that by converting "no timeout" to "wait for one year"